File: src/main/java/ca/yyx/hu/aap/AapAudio.java
package ca.yyx.hu.aap;

import android.media.AudioManager;
import android.util.SparseIntArray;

import ca.yyx.hu.aap.protocol.AudioConfigs;
import ca.yyx.hu.aap.protocol.Channel;
import ca.yyx.hu.aap.protocol.MsgType;
import ca.yyx.hu.aap.protocol.nano.Protocol;
import ca.yyx.hu.decoder.AudioDecoder;
import ca.yyx.hu.utils.AppLog;

/**
 * @author algavris
 * @date 01/10/2016.
 */

class AapAudio implements AudioManager.OnAudioFocusChangeListener {
    private final AudioDecoder mAudioDecoder;

    private static final int AUDIO_BUFS_SIZE = 65536 * 4;      // Up to 256 Kbytes
    private final AudioManager mAudioManager;

    AapAudio(AudioDecoder audioDecoder, AudioManager audioManager) {
        mAudioDecoder = audioDecoder;
        mAudioManager = audioManager;
        mAudioManager.requestAudioFocus(this, AudioManager.STREAM_MUSIC, AudioManager.AUDIOFOCUS_GAIN);
    }

    void requestFocusChange(int focusRequest)
    {
        int stream = AudioManager.STREAM_MUSIC;
        if (focusRequest == Protocol.AudioFocusRequestNotification.AUDIO_FOCUS_RELEASE) {
            mAudioManager.abandonAudioFocus(this);
        } else if (focusRequest == Protocol.AudioFocusRequestNotification.AUDIO_FOCUS_GAIN) {
            mAudioManager.requestAudioFocus(this, stream, AudioManager.AUDIOFOCUS_GAIN);
        } else if (focusRequest == Protocol.AudioFocusRequestNotification.AUDIO_FOCUS_GAIN_TRANSIENT) {
            mAudioManager.requestAudioFocus(this, stream, AudioManager.AUDIOFOCUS_GAIN_TRANSIENT);
        } else if (focusRequest == Protocol.AudioFocusRequestNotification.AUDIO_FOCUS_UNKNOWN) {
            mAudioManager.requestAudioFocus(this, stream, AudioManager.AUDIOFOCUS_GAIN_TRANSIENT_MAY_DUCK);
        }
    }

    public int process(AapMessage message) {
        if (message.length >= 10) {
            decode(message.channel, 10, message.data, message.length - 10);
        }

        return 0;
    }

    private void decode(int channel, int start, byte[] buf, int len) {
        if (len > AUDIO_BUFS_SIZE) {
            AppLog.e("Error audio len: %d  aud_buf_BUFS_SIZE: %d", len, AUDIO_BUFS_SIZE);
            len = AUDIO_BUFS_SIZE;
        }

        if (mAudioDecoder.getTrack(channel) == null)
        {
            Protocol.AudioConfiguration config = AudioConfigs.get(channel);
            int stream = AudioManager.STREAM_MUSIC;
            mAudioDecoder.start(channel, stream, config.sampleRate, config.numberOfBits, config.numberOfChannels);
        }

        mAudioDecoder.decode(channel, buf, start, len);
    }

    void stopAudio(int chan) {
        AppLog.i("Audio Stop: " + chan);
        mAudioDecoder.stop(chan);
    }

    @Override
    public void onAudioFocusChange(int focusChange) {
        AppLog.i("" + focusChange);
    }
}

File: src/main/java/ca/yyx/hu/aap/AapAudio.kt
package ca.yyx.hu.aap

import android.media.AudioManager

import ca.yyx.hu.aap.protocol.AudioConfigs
import ca.yyx.hu.aap.protocol.Channel
import ca.yyx.hu.aap.protocol.nano.Control
import ca.yyx.hu.decoder.AudioDecoder
import ca.yyx.hu.utils.AppLog

/**
 * @author algavris
 * *
 * @date 01/10/2016.
 * *
 * *
 * @link https://github.com/google/ExoPlayer/blob/release-v2/library/src/main/java/com/google/android/exoplayer2/audio/AudioTrack.java
 */
internal class AapAudio(
        private val audioDecoder: AudioDecoder,
        private val audioManager: AudioManager) {

    fun requestFocusChange(stream: Int, focusRequest: Int, callback: AudioManager.OnAudioFocusChangeListener) {
        when (focusRequest) {
            Control.AudioFocusRequestNotification.AUDIOFOCUS_RELEASE -> audioManager.abandonAudioFocus(callback)
            Control.AudioFocusRequestNotification.AUDIOFOCUS_GAIN -> audioManager.requestAudioFocus(callback, stream, AudioManager.AUDIOFOCUS_GAIN)
            Control.AudioFocusRequestNotification.AUDIOFOCUS_GAIN_TRANSIENT -> audioManager.requestAudioFocus(callback, stream, AudioManager.AUDIOFOCUS_GAIN_TRANSIENT)
            Control.AudioFocusRequestNotification.AUDIOFOCUS_GAIN_TRANSIENT_MAY_DUCK -> audioManager.requestAudioFocus(callback, stream, AudioManager.AUDIOFOCUS_GAIN_TRANSIENT_MAY_DUCK)
        }
    }

    fun process(message: AapMessage): Int {
        if (message.size >= 10) {
            decode(message.channel, 10, message.data, message.size - 10)
        }

        return 0
    }

    private fun decode(channel: Int, start: Int, buf: ByteArray, len: Int) {
        var length = len
        if (length > AUDIO_BUFS_SIZE) {
            AppLog.e("Error audio len: %d  aud_buf_BUFS_SIZE: %d", length, AUDIO_BUFS_SIZE)
            length = AUDIO_BUFS_SIZE
        }

        if (audioDecoder.getTrack(channel) == null) {
            val config = AudioConfigs.get(channel)
            val stream = AudioManager.STREAM_MUSIC
            audioDecoder.start(channel, stream, config.sampleRate, config.numberOfBits, config.numberOfChannels)
        }

        audioDecoder.decode(channel, buf, start, length)
    }

    fun stopAudio(channel: Int) {
        AppLog.i("Audio Stop: " + Channel.name(channel))
        audioDecoder.stop(channel)
    }

    companion object {
        private const val AUDIO_BUFS_SIZE = 65536 * 4  // Up to 256 Kbytes
    }
}

